src/transformers/models/lightglue/modeling_lightglue.py

```diff
@@ -0,0 +1,1167 @@
+        m_descriptors = m_descriptors / self.descriptor_dim**0.25
+        matchability = self.matchability(descriptors)
+        batch_size, num_keypoints, descriptor_dim = descriptors.shape
+        m_descriptors = m_descriptors.reshape(batch_size // 2, 2, num_keypoints, descriptor_dim)
+        m_descriptors0 = m_descriptors[:, 0]
+        m_descriptors1 = m_descriptors[:, 1]
+        similarity = torch.einsum("bmd,bnd->bmn", m_descriptors0, m_descriptors1)
+        if mask is not None:
+            mask = mask.reshape(batch_size // 2, 2, num_keypoints)
+            mask0 = mask[:, 0].unsqueeze(-1)
+            mask1 = mask[:, 1].unsqueeze(-1).transpose(-1, -2)
+            mask = mask0 * mask1
+            similarity = similarity.masked_fill(mask == 0, torch.finfo(similarity.dtype).min)
+
+        matchability = matchability.reshape(batch_size // 2, 2, num_keypoints, 1)
+        matchability_0 = matchability[:, 0]
+        matchability_1 = matchability[:, 1]
+        scores = sigmoid_log_double_softmax(similarity, matchability_0, matchability_1)
+        return scores
+
+    def get_matchability(self, descriptors: torch.Tensor) -> torch.Tensor:
+        matchability = self.matchability(descriptors)
+        matchability = nn.functional.sigmoid(matchability).squeeze(-1)
+        return matchability
+
+
+class LightGlueTokenConfidenceLayer(nn.Module):
+    def __init__(self, config: LightGlueConfig):
+        super().__init__()
+
+        self.token = nn.Linear(config.descriptor_dim, 1)
+
+    def forward(self, descriptors: torch.Tensor) -> torch.Tensor:
+        token = self.token(descriptors.detach())
+        token = nn.functional.sigmoid(token).squeeze(-1)
+        return token
+
+
+def filter_matches(scores: torch.Tensor, threshold: float) -> Tuple[torch.Tensor, torch.Tensor]:
+    """obtain matches from a log assignment matrix [Bx M+1 x N+1]"""
+    batch_size, _, _ = scores.shape
+    max0, max1 = scores[:, :-1, :-1].max(2), scores[:, :-1, :-1].max(1)
+    matches0, matches1 = max0.indices, max1.indices
+    indices0 = torch.arange(matches0.shape[1], device=matches0.device)[None]
+    indices1 = torch.arange(matches1.shape[1], device=matches1.device)[None]
+    mutual0 = indices0 == matches1.gather(1, matches0)
+    mutual1 = indices1 == matches0.gather(1, matches1)
+    max0 = max0.values.exp()
+    zero = max0.new_tensor(0)
+    matching_scores0 = torch.where(mutual0, max0, zero)
+    matching_scores1 = torch.where(mutual1, matching_scores0.gather(1, matches1), zero)
+    valid0 = mutual0 & (matching_scores0 > threshold)
+    valid1 = mutual1 & valid0.gather(1, matches1)
+    matches0 = torch.where(valid0, matches0, -1)
+    matches1 = torch.where(valid1, matches1, -1)
+    matches = torch.stack([matches0, matches1]).transpose(0, 1).reshape(batch_size * 2, -1)
+    matching_scores = torch.stack([matching_scores0, matching_scores1]).transpose(0, 1).reshape(batch_size * 2, -1)
+
+    return matches, matching_scores
```

Comment:
Also looks like lots code lines, would be nice to add some docs, comments

Reply:
Added comments to this function for more clarity